rust-mcd/src/connection.rs

```diff
@@ -0,0 +1,102 @@
+use std::{ffi::CStr, fmt::Debug};
+
+use crate::{error::McdError, mcd_bindings::mcd_server_info_st, system::System, MCD_LIB};
+
+pub struct Connection {
+    servers: Vec<mcd_server_info_st>,
+}
+
+impl Connection {
+    /// Scan for available servers
+    pub fn scan() -> anyhow::Result<Self> {
+        let host = CStr::from_bytes_with_nul(b"localhost\0").unwrap();
+        let server_count = MCD_LIB.query_server_count(host)?;
+        let servers = MCD_LIB
+            .query_server_infos(host, server_count)
+            .add_mcd_error_info(None)?;
+
+        let connection = Connection { servers };
+
+        log::trace!("Scanned for servers, found {connection:?}");
+
+        Ok(connection)
+    }
+
+    /// List all servers available in this connection
+    pub fn servers(&self) -> impl Iterator<Item = ServerInfo> + '_ {
+        self.servers.iter().map(ServerInfo::from)
+    }
+
+    /// Number of servers available
+    pub fn count(&self) -> usize {
+        self.servers.len()
+    }
+}
+
+/// Information a server
+#[derive(Clone, Copy)]
+pub struct ServerInfo {
+    pub(crate) inner: mcd_server_info_st,
+}
```

Comment:
You might want to clarify, I am not sure if I fully understood.

I assume that you suggest making `ServerInfo` a transparent structure, so we can transmute from `&mcd_server_info_st` to `&ServerInfo`? I can see there are some benefits, but I am not sure what `unsafe` code this would avoid.

Reply:
Still not a 100% sure what you mean here, lets chat off-github tomorrow.